### PR TITLE
optimize: Optimization for secret retrieving logic

### LINKED ISSUE #3742: Optimization for secret retrieving logic
## Feature Request

**Is your feature request related to a problem? Please describe:**

**Describe the feature you'd like:**

Currently, the secrets are retrieved from the kube-apiserver directly in many functions, which will increase the load of kube-apiserver, and should be optimized to use the secretLister.
```
pkg/pdapi/pdapi.go:	secret, err := kubeCli.CoreV1().Secrets(string(namespace)).Get(secretName, types.GetOptions{})
pkg/controller/http_client.go:	secret, err := c.kubeCli.CoreV1().Secrets(ns).Get(secretName, metav1.GetOptions{})
pkg/autoscaler/autoscaler/query/external.go:	secret, err := kubecli.CoreV1().Secrets(endpoint.TLSSecret.Namespace).Get(endpoint.TLSSecret.Name, metav1.GetOptions{})
pkg/backup/util/util.go:			secret, err := kubeCli.CoreV1().Secrets(ns).Get(s3SecretName, metav1.GetOptions{})
pkg/backup/util/util.go:			secret, err := kubeCli.CoreV1().Secrets(ns).Get(gcsSecretName, metav1.GetOptions{})
pkg/backup/util/util.go:	secret, err := kubeCli.CoreV1().Secrets(ns).Get(tidbSecretName, metav1.GetOptions{})
```
**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by mianhk**
### What problem does this PR solve?

close #3742 
### Code changes

- [ ] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [x] Unit test 
- [x] E2E test 
- [ ] Manual test 
- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Optimization for secret retrieving logic
```

### CODE REVIEW

File: pkg/autoscaler/autoscaler/query/external.go [hunk 1]
```diff
@@ -104,8 +102,8 @@ func getClient(endpoint v1alpha1.ExternalEndpoint, kubecli kubernetes.Interface)
 	return client, nil
 }
 
-func loadTLSConfig(endpoint v1alpha1.ExternalEndpoint, kubecli kubernetes.Interface) (*tls.Config, error) {
-	secret, err := kubecli.CoreV1().Secrets(endpoint.TLSSecret.Namespace).Get(context.Background(), endpoint.TLSSecret.Name, metav1.GetOptions{})
+func loadTLSConfig(endpoint v1alpha1.ExternalEndpoint, secretLister corelisterv1.SecretLister) (*tls.Config, error) {
+	secret, err := secretLister.Secrets(string(endpoint.TLSSecret.Namespace)).Get(endpoint.TLSSecret.Name)
```
[the remark this answers is not in the document]
Comment (mianhk) [reply pkg/autoscaler/autoscaler/query/external.go]: Yes, fixed 😄

**Review by csuzhangxc — APPROVED** [review]

**Review by july2993 — APPROVED** [review]

File: cmd/discovery/main.go [hunk 2]
```diff
@@ -87,11 +90,25 @@ func main() {
 	if tlsEnabled == strconv.FormatBool(true) {
 		tcTls = true
 	}
+	// informers
+	options := []informers.SharedInformerOption{
+		informers.WithNamespace(os.Getenv("MY_POD_NAMESPACE")),
+	}
+
+	kubeInformerFactory := kubeinformers.NewSharedInformerFactoryWithOptions(kubeCli, 30*time.Second, options...)
```
Comment (DanielZhangQD) [inline cmd/discovery/main.go:98]:
```suggestion
	kubeInformerFactory := kubeinformers.NewSharedInformerFactoryWithOptions(kubeCli, 30*time.Minute, options...)
```

Comment (DanielZhangQD) [reply cmd/discovery/main.go]: No need to resync too frequently.

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Pull request merged**